### PR TITLE
DRY up ContentBlocksList and BlockInspectorLockedBlocks

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Follows https://github.com/WordPress/gutenberg/pull/50857.

Extract a reusable component out of `ContentBlocksList` and `BlockInspectorLockedBlocks` so that we're not repeating the same code that outputs a list of block navigation links.

## Why?
Less code good.

## Testing Instructions
1. Go to Appearance → Editor → Pages and select a page.
1. The right sidebar should display a list containing Post Featured Image, Title, and Post Content. Clicking on items in this list should select the block. The items should appear selected if the block or one of its children are selected.
1. Create a new post and insert a pattern that contains a `contentOnly` lock. ([Here's one.](https://gist.github.com/richtabor/ddeea41ced691721318649bea8ce9db8))
1. Select the pattern.
1. The right sidebar should display a list containing the editable blocks within that pattern. Clicking on items in this list should select the block. The items should appear selected if the block is selected.

### CODE REVIEW

File: packages/block-editor/src/components/block-quick-navigation/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { getBlockType, __experimentalGetBlockLabel } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import BlockIcon from '../block-icon';
+import { unlock } from '../../lock-unlock';
+
+export default function BlockQuickNavigation( { clientIds } ) {
```
Comment (noisysocks) [inline packages/block-editor/src/components/block-quick-navigation/index.js:19]: Not certain about what to name this component.

File: packages/block-editor/src/components/block-inspector/index.js [hunk 2]
```diff
@@ -38,77 +29,26 @@ import AdvancedControls from '../inspector-controls-tabs/advanced-controls-panel
...
-					topLevelLockedBlock
-				),
-				selectedBlock: select( blockEditorStore ).getSelectedBlock(),
-			};
+			const {
+				getClientIdsOfDescendants,
+				getBlockName,
+				getBlockEditingMode,
+			} = unlock( select( blockEditorStore ) );
+			return getClientIdsOfDescendants( [ topLevelLockedBlock ] ).filter(
```
Comment (Mamaduka) [inline packages/block-editor/src/components/block-inspector/index.js:43]: This is a nice trick to get around the array/object referential equality "requirement" and avoid excessive re-renders 🥇

**Comment by apeatling** [comment]
I followed the steps and inserted the pattern. Selected it, but I still don't see the editable blocks in the right sidebar. Maybe I'm missing a step. Is there a specific way the pattern should be inserted?

Screenshot 2023-06-07 at 10 17 44 AM

**Comment by noisysocks** [comment]
@apeatling: It should appear in the Block tab. Probably easiest to test patterns while editing a template or using the post editor.

File: packages/block-editor/src/components/block-inspector/index.js (hunk 2 above)
Comment (noisysocks) [reply packages/block-editor/src/components/block-inspector/index.js:43]: Wait... it is? Won't `filter` return a new array every time? 😅

Comment (noisysocks) [reply packages/block-editor/src/components/block-inspector/index.js:43]: You sent me down a rabbit hole of optimising all of this. I made some progress but will tackle it in a separate PR 😅

Comment (Mamaduka) [reply packages/block-editor/src/components/block-inspector/index.js:43]: Sorry about that 😅 I left two more comments in other PR, where returned values need optimization.

This map select returns one dimensional array of client IDs, correct?

Comment (noisysocks) [reply packages/block-editor/src/components/block-inspector/index.js:43]: This comment? https://github.com/WordPress/gutenberg/pull/50643#discussion_r1210261515 Yes I've been looking into that.

Yes this is a one dimensional array.

Comment (noisysocks) [reply packages/block-editor/src/components/block-inspector/index.js:43]: I pushed 6006434b8b13f5d214f9b05c35b8dd96748e971a which reduces the number of re-renders in `PagePanels` and `BlockQuickNavigation` by selecting the minimal amount of data. I think `BlockInspector` here is as good as we can get it without creating a new selector which is not worth it right now imo.

I have more performance improvements ready but they're unrelated so they'll be in a separate PR 😀

Comment (noisysocks) [reply packages/block-editor/src/components/block-inspector/index.js:43]: Follow-up: https://github.com/WordPress/gutenberg/pull/51319

**Review by apeatling — APPROVED** [review]
Tried this again and can confirm it works as described. 👍

File: packages/block-editor/src/store/selectors.js [hunk 3]
```diff
@@ -303,10 +303,13 @@ export const __experimentalGetGlobalBlocksByName = createSelector(
...
 			return EMPTY_ARRAY;
 		}
+		const blockNames = Array.isArray( blockName )
+			? blockName
+			: [ blockName ];
 		const clientIds = getClientIdsWithDescendants( state );
 		const foundBlocks = clientIds.filter( ( clientId ) => {
 			const block = state.blocks.byClientId.get( clientId );
-			return block.name === blockName;
+			return blockNames.includes( block.name );
```
Comment (Mamaduka) [inline packages/block-editor/src/store/selectors.js:312]: @noisysocks, do you remember why you introduced this change? I can't find any place that passes the block names array to this selector.

**Pull request merged**